src/main/scala/strawman/collection/immutable/Range.scala

```diff
@@ -0,0 +1,280 @@
+package strawman.collection.immutable
+
+import strawman.collection.mutable.Iterator
+import strawman.collection.{IndexedSeq, Iterable, IterableFactory, IterableOnce, IterablePolyTransformsFromIterable, Seq, SeqLike, SeqMonoTransformsFromIterable}
+
+import scala.{Boolean, IllegalArgumentException, IndexOutOfBoundsException, Int, inline, Long, StringContext, Serializable, SerialVersionUID, Unit}
+import scala.Predef.{???, augmentString}
+import java.lang.String
+
+/** The `Range` class represents integer values in range
+  *  ''[start;end)'' with non-zero step value `step`.
+  *  It's a special case of an indexed sequence.
+  *  For example:
+  *
+  *  {{{
+  *     val r1 = 0 until 10
+  *     val r2 = r1.start until r1.end by r1.step + 1
+  *     println(r2.length) // = 5
+  *  }}}
+  *
+  *  Ranges that contain more than `Int.MaxValue` elements can be created, but
+  *  these overfull ranges have only limited capabilities. Any method that
+  *  could require a collection of over `Int.MaxValue` length to be created, or
+  *  could be asked to index beyond `Int.MaxValue` elements will throw an
+  *  exception. Overfull ranges can safely be reduced in size by changing
+  *  the step size (e.g. `by 3`) or taking/dropping elements. `contains`,
+  *  `equals`, and access to the ends of the range (`head`, `last`, `tail`,
+  *  `init`) are also permitted on overfull ranges.
+  *
+  *  @param start       the start of this range.
+  *  @param end         the end of the range.  For exclusive ranges, e.g.
+  *                     `Range(0,3)` or `(0 until 3)`, this is one
+  *                     step past the last one in the range.  For inclusive
+  *                     ranges, e.g. `Range.inclusive(0,3)` or `(0 to 3)`,
+  *                     it may be in the range if it is not skipped by the step size.
+  *                     To find the last element inside a non-empty range,
+  *                     use `last` instead.
+  *  @param step        the step for the range.
+  *  @param isInclusive whether the end of the range is included or not
+  */
+@SerialVersionUID(7618862778670199309L)
+final class Range(
+  val start: Int,
+  val end: Int,
+  val step: Int,
+  val isInclusive: Boolean
+)
+  extends IndexedSeq[Int]
+    with IterableFactory[IndexedSeq]
+    with IterablePolyTransformsFromIterable[Int, IndexedSeq]
+    with SeqMonoTransformsFromIterable[Int, IndexedSeq[Int]]
+    with Serializable { range =>
+
+  def iterator(): Iterator[Int] =
+    new Iterator[Int] {
+      private var current: Seq[Int] = range
+      def next() = {
+        val x = current.head
+        current = current.tail
```

Comment:
We can do *much* better than `current.tail`. If `current` was simply an `Int`, it's `current += step`. This is important because create a `Range` has a non-trivial cost. Iterating over a `Range` should not have to allocate $n$ instances of `Range`!